src/com/facebook/buck/intellij/ideabuck/src/com/facebook/buck/intellij/ideabuck/actions/select/RunSelectedTestAction.java

```diff
@@ -0,0 +1,172 @@
+    if (buckFile != null) {
+      String name;
+      Optional<String> testSelectors = Optional.empty();
+      if (method != null) {
+        name = psiClass.getName() + "#" + method.getName();
+        testSelectors = Optional.of(psiClass.getName() + "#" + method.getName());
+      } else {
+        name = element.getContainingFile().getName();
+        if (element instanceof PsiClass) {
+          testSelectors = Optional.of(psiClass.getQualifiedName());
+        }
+      }
+      createTestConfigurationFromContext(
+          name,
+          testSelectors,
+          project,
+          virtualFile,
+          buckFile,
+          debug);
+    }
+  }
+
+  /**
+   * Create a test Configuration and run or debug it in the ide.
+   *
+   * @param name a {@link String} representing the name of the configuration.
+   * @param testSelectors a {@link String} representing optional testSelectors for filtering.
+   * @param project {@link Project} then intellij project corresponding the the file or module under test.
+   * @param containingFile {@link VirtualFile} the file that containing the impacted tests.
+   * @param buckFile {@link VirtualFile} the file representing the buck File.
+   * @param debug a {@link boolean} used to signify run or debug.
+   */
+  private void createTestConfigurationFromContext(
+      @Nonnull String name,
+      final Optional<String> testSelectors,
+      final Project project,
+      VirtualFile containingFile,
+      VirtualFile buckFile,
+      boolean debug) {
+    BuckBuildManager buildManager = BuckBuildManager.getInstance(project);
+    BuckQueryCommandHandler handler =
+      new BuckQueryCommandHandler(
+          project,
+          buckFile.getParent(),
+          BuckCommand.QUERY, new Function<List<String>, Void>() {
+        @Nullable
+        @Override
+        public Void apply(@Nullable List<String> strings) {
+          RunManagerImpl runManager = (RunManagerImpl) RunManager.getInstance(project);
+          TestConfigurationType type = new TestConfigurationType();
+          RunnerAndConfigurationSettingsImpl runnerAndConfigurationSettings
+              = (RunnerAndConfigurationSettingsImpl)runManager.createRunConfiguration(name, type.getConfigurationFactories()[0]);
+          TestConfiguration testConfiguration = (TestConfiguration) runnerAndConfigurationSettings.getConfiguration();
+          testConfiguration.data.target = strings.get(0);
+          testConfiguration.data.testSelectors = testSelectors.isPresent() ? testSelectors.get() : "";
+          runnerAndConfigurationSettingsResult = runnerAndConfigurationSettings;
+          runManager.addConfiguration(runnerAndConfigurationSettings, false);
+          return null;
+        }});
+    handler.command().addParameter("owner("+ containingFile.getPath()+")");
+    buildManager.runInCurrentThreadPostEnd(handler,
+      () -> {
+        if (runnerAndConfigurationSettingsResult != null) {
```

Comment:
This one was actually a pain. Intellij is so picky about which thread it will let you set configurations on.